### PR TITLE
inductor cpp wrapper: add GIL release and acquire

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111888

Support multiple instances inference (in different threads of the same process) as in https://github.com/pytorch/pytorch/issues/93524#issuecomment-1421816158.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -1427,6 +1433,7 @@ def generate_return(self, output_refs):
                         + f"new at::Tensor({output}));"
                     )
         else:
+            self.wrapper_call.writeline("py::gil_scoped_acquire acquire;")
```
Comment (jgong5) [inline torch/_inductor/codegen/wrapper.py:1436]: with scoped release at the beginning of the function, you don't need to explicitly acquire the gil here, right?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/wrapper.py]: Removed the explicit acquire here.

**Review by jgong5 — APPROVED** [review]

**Review by desertfire — COMMENTED** [review]
Test case?

**Review by jansel — COMMENTED** [review]
Is it currently possible for cpp wrapper to call into python code?

**Comment by jgong5** [comment]
> Is it currently possible for cpp wrapper to call into python code?

I'm not aware of a path in cpp wrapper to call into python code? Any concern about it?

**Comment by chunyuan-w** [comment]
> Test case?

I added a test case to cover the functionality of a multi-threading scenario.

**Review by jansel — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

**Pull request closed without merging**